addons/test_mail/tests/test_mail_multicompany.py

```diff
@@ -253,62 +255,156 @@ def test_post_wo_access(self):
+                            "activity_ids": set(actual_activity_group['activity_ids']),
+                        }
+                    )
+                test_model_name = 'mail.test.multi.company.with.activity'
+                self.assertIn(test_model_name, actual_activity_groups_by_model)
+                self.assertDictEqual(
+                    {
+                        "type": "activity",
+                        "view_type": "list",
+                        "overdue_count": 0,
+                        "planned_count": 0,
+                        "today_count": len(expected_test_model_activities),
+                        "total_count": len(expected_test_model_activities),
+                        "id": self.env["ir.model"]._get_id(test_model_name),
+                        "model": test_model_name,
+                        "name": "Test Multi Company Mail With Activity",
+                        "icon": "/base/static/description/icon.png",
+                    },
+                    actual_activity_groups_by_model[test_model_name])
+                allowed_user_company_records = _mock_filter_access_rules(
+                    (self.env[test_model_name]
+                    ).with_context(allowed_company_ids=allowed_company_ids).with_user(user).search([
+                        ('id', 'in', test_records_all.ids),
+                        ('activity_user_id', '=', user.id),
+                    ]),
+                    'read')
+                self.assertEqual(
+                    (self.env['mail.activity']
+                    ).with_user(user).with_context(allowed_company_ids=allowed_company_ids).search([
+                        ('res_model', '=', test_model_name),
+                        ('res_id', 'in', allowed_user_company_records.ids),
+                        ('user_id', '=', user.id),
+                    ]),
+                    expected_test_model_activities)
+                self.assertEqual(len(allowed_user_company_records), len(expected_test_model_activities),
+                                 'One scheduled activity per record per user')
+        # Employee cannot fetch activities of a company he doesn't belong to
+        self.authenticate(user_employee.login, user_employee.login)
+        with self.assertRaises(JsonRpcException, msg="odoo.exceptions.AccessError"), mute_logger("odoo.http"):
+            self.make_jsonrpc_request("/mail/data", {
+                "systray_get_activities": True,
+                "context": {"allowed_company_ids": company_2_admin_only.ids}})
+        # Activities related to not accessible records are in other activities regardless of the allowed companies
+        self.authenticate(user_admin.login, user_admin.login)
+        with patch.object(MailTestMultiCompanyWithActivity, 'check_access_rights', autospec=True,
+                          side_effect=lambda *args, **kwargs: False):
+            for company in (company_1_all, company_2_admin_only):
```

Comment:
Could add a test case with both companies to be complete.

Also, could benefit from subTest with the company as argument.